tests/layer/test_graph_attention.py

```diff
@@ -0,0 +1,375 @@
+    G.add_edges_from(elist, label="default")
+
+    # Add example features
+    if feature_size is not None:
+        for v in G.nodes():
+            G.node[v]["feature"] = np.ones(feature_size)
+        return StellarGraph(G, node_features="feature")
+
+    else:
+        return StellarGraph(G)
+
+
+class Test_GraphAttention_layer:
+    N = 10
+    F_in = 5
+    F_out = 2
+    attn_heads = 8
+    activation = "relu"
+    """
+    Tests of GraphAttention layer
+    """
+
+    def test_constructor(self):
+        # attn_heads_reduction = "concat":
+        layer = GraphAttention(
+            F_out=self.F_out,
+            attn_heads=self.attn_heads,
+            attn_heads_reduction="concat",
+            activation=self.activation,
+        )
+        assert layer.F_out == self.F_out
+        assert layer.attn_heads == self.attn_heads
+        assert layer.output_dim == self.F_out * self.attn_heads
+        assert layer.activation == keras.activations.get(self.activation)
+
+        # attn_heads_reduction = "average":
+        layer = GraphAttention(
+            F_out=self.F_out,
+            attn_heads=self.attn_heads,
+            attn_heads_reduction="average",
+            activation=self.activation,
+        )
+        assert layer.output_dim == self.F_out
+
+        # attn_heads_reduction = "ave":
+        with pytest.raises(ValueError):
+            GraphAttention(
+                F_out=self.F_out,
+                attn_heads=self.attn_heads,
+                attn_heads_reduction="ave",
+                activation=self.activation,
+            )
+
+    def test_apply_concat(self):
+        gat = GraphAttention(
+            F_out=self.F_out,
+            attn_heads=self.attn_heads,
+            attn_heads_reduction="concat",
+            activation=self.activation,
+            kernel_initializer="ones",
+        )
+        x_inp = [Input(shape=(self.F_in,)), Input(shape=(self.N,))]
+        x_out = gat(x_inp)
+
+        model = keras.Model(inputs=x_inp, outputs=x_out)
+        assert model.output_shape[-1] == self.F_out * self.attn_heads
+
+        X = np.ones((self.N, self.F_in))  # features
+        A = np.eye(self.N)  # adjacency matrix with self-loops only
+        expected = np.ones((self.N, self.F_out * self.attn_heads)) * self.F_in
+        actual = model.predict([X, A])
+        assert expected == pytest.approx(actual)
+
+    def test_apply_average(self):
+        gat = GraphAttention(
+            F_out=self.F_out,
+            attn_heads=self.attn_heads,
+            attn_heads_reduction="average",
+            activation=self.activation,
+            kernel_initializer="ones",
+        )
+        x_inp = [Input(shape=(self.F_in,)), Input(shape=(self.N,))]
+        x_out = gat(x_inp)
+
+        model = keras.Model(inputs=x_inp, outputs=x_out)
+        assert model.output_shape[-1] == self.F_out
+
+        X = np.ones((self.N, self.F_in))  # features
+        A = np.eye(self.N)  # adjacency matrix with self-loops only
```

Comment:
Perhaps a test with a non-identity adjacency matrix would be useful?

Reply:
what would that test check?

Reply:
Check the application of the attention layer to a graph with some neighbours.

Reply:
added one such (simple) test with nodes 0-1 connected